### ISSUE TITLE
Boundary options seem to be ignored

### ISSUE THREAD

**Issue by Denn0** [weaveworks/eksctl#4764]
The following configuration fails because of our AWS Service Control Policies that state that all roles created should have a boundary policy attached. We seem to be providing boundary policies, but `eksctl` still seems to be creating roles without boundaries, which is denied based on the mentioned SCPs. 
`eksctl` should use the provided boundary policies in ALL IAM roles it tries to create.

```code:yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: datahubproject
  region: eu-central-1

vpc:
  subnets:
    private:
      eu-central-1a: { id: subnet-123 }
      eu-central-1c: { id: subnet-456 }
      eu-central-1b: { id: subnet-789 }
  clusterEndpoints:
    privateAccess: true
    publicAccess: false

iam:
  withOIDC: true
  serviceRolePermissionsBoundary: "arn:aws:iam::000000000000:policy/boundary"
  fargatePodExecutionRolePermissionsBoundary: "arn:aws:iam::000000000000:policy/boundary"
  serviceAccounts:
    - metadata:
        name: datahubproject-role
      attachPolicyARNs:
      - "arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess"
      permissionsBoundary: "arn:aws:iam::000000000000:policy/boundary"

nodeGroups:
  - name: ng-1-workers
    labels: { role: workers }
    instanceType: m5.medium
    desiredCapacity: 3
    privateNetworking: true
    iam:
      instanceRolePermissionsBoundary: "arn:aws:iam::000000000000:policy/boundary"
```

Fails on this CloudFormation template it tries to create as it is missing the boundary role: 
```code:json
{
    "AWSTemplateFormatVersion": "2010-09-09",
    "Description": "IAM role for serviceaccount \"kube-system/aws-node\" [created and managed by eksctl]",
    "Resources": {
        "Role1": {
            "Type": "AWS::IAM::Role",
            "Properties": {
                "AssumeRolePolicyDocument": {
                    "Statement": [{
                            "Action": ["sts:AssumeRoleWithWebIdentity"],
                            "Condition": {
                                "StringEquals": {
                                    "oidc.eks.eu-central-1.amazonaws.com/id/9A6001CB878A68BAA71F00A1234076F6:aud": "sts.amazonaws.com",
                                    "oidc.eks.eu-central-1.amazonaws.com/id/9A6001CB878A68BAA71F00A1234076F6:sub": "system:serviceaccount:kube-system:aws-node"
                                }
                            },
                            "Effect": "Allow",
                            "Principal": {
                                "Federated": "arn:aws:iam::000000000000:oidc-provider/oidc.eks.eu-central-1.amazonaws.com/id/9A6001CB878A68BAA71F00A1234076F6"
                            }
                        }
                    ],
                    "Version": "2012-10-17"
                },
                "ManagedPolicyArns": ["arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy"]
            }
        }
    },
    "Outputs": {
        "Role1": {
            "Value": {
                "Fn::GetAtt": "Role1.Arn"
            }
        }
    }
}
```

There's also no workaround to fix the template, and then continue the process unfortunately.
So I'm forced to manually create the whole setup from scratch now, basically rendering `ekscl` useless in our case. 

Am I missing some option in the config here? 

Thanks! 

### Summary
- We need to do some investigation as part of this ticket to find out what's failing
- Might help if the user provides the commands they're using as it's not bit clear how they are running this exactly.
- Timeboxed to half a day to reproduce this.

**Comment by Himangini** [comment]
@Denn0 Thanks for reporting this bug 👍 If you have a fix in mind, we are happy to accept PRs, otherwise the core team will get to it soon

**Comment by aclevername** [comment]
I've been investigating and believe I understand the issue. When you create a cluster with `withOIDC: true`  we automatically (not visible to the user) create an additional IAMServiceAccount for the VPC-CNI. This is because this approach is much better and more secure than giving the VPC-CNI permissions to the node IAM role. Docs [here](https://eksctl.io/usage/security/#withoidc) and [here](https://docs.aws.amazon.com/eks/latest/userguide/cni-iam-role.html).

The outcome of the fact we automatically add this IAMServiceAccount is that the user doesn't know we're adding it, and therefore can't choose to specify a `permissionBoundary` for it. We have two possible outcomes:

1. Leave this functionality as is. Update the docs to make it clear that this is the behaviour, and document that if you want to add a permission boundary to the VPC-CNI then you must manually add the VPC-CNI IAMServiceAccount to your config file with the field set (which prevents eksctl from attempting to add it):
```
iam:
  serviceAccounts:
    - metadata:
        name: aws-node
      attachPolicyARNs:
      - "arn:aws:iam::000000000000:policy/AmazonEKS_CNI_Policy"
      permissionsBoundary: "arn:aws:iam::000000000000:policy/boundary-you-care-about"
```

2. Introduce a new field to `iam` section that the user can configure to let us know to add a permission boundary to the VPC-CNI. This is a bit ugly, since its 1 top-level field for an IAMServiceAccount the user can't even see :/ 

I think option 1 makes more sense, and can be done now as a workaround. If we decide to do option 2 we can always update the docs to reflect this.

@Denn0 can you try this workaround and confirm it works for you? Thanks :smile:

**Comment by Himangini** [comment]
> @Denn0 can you try this workaround and confirm it works for you?
@Denn0 any updates on this, please?

**Comment by Denn0** [comment]
Hi, 

Sorry for not responding. Actually we moved on to using CDK to deploy our EKS clusters, it gave us a bit more control. 
I just raised this issue as the docs seem to say that if I provide a permission boundary (it's not specific on which field exactly) 

> all identity-based entities created by eksctl are created within that boundary
https://eksctl.io/usage/iam-permissions-boundary/

If you're looking for a way to solve this; I would have interpreted the docs that if I set `serviceRolePermissionsBoundary`, all roles created will get that boundary. That would work just fine for me. 

But like I said, we skipped using eksctl for now.

Thanks guys!
